cpp/ql/lib/experimental/semmle/code/cpp/dataflow/ProductFlow.qll

```diff
@@ -324,25 +324,71 @@ module ProductFlow {
 
     module Flow2 = DataFlow::GlobalWithState<Config2>;
 
+    private predicate isSourcePair(Flow1::PathNode node1, Flow2::PathNode node2) {
+      Config::isSourcePair(node1.getNode(), node1.getState(), node2.getNode(), node2.getState())
+    }
+
+    private predicate isSinkPair(Flow1::PathNode node1, Flow2::PathNode node2) {
+      Config::isSinkPair(node1.getNode(), node1.getState(), node2.getNode(), node2.getState())
+    }
+
+    pragma[assume_small_delta]
     pragma[nomagic]
-    private predicate reachableInterprocEntry(
-      Flow1::PathNode source1, Flow2::PathNode source2, Flow1::PathNode node1, Flow2::PathNode node2
-    ) {
-      Config::isSourcePair(node1.getNode(), node1.getState(), node2.getNode(), node2.getState()) and
-      node1 = source1 and
-      node2 = source2
+    private predicate fwdReachableInterprocEntry(Flow1::PathNode node1, Flow2::PathNode node2) {
+      isSourcePair(node1, node2)
       or
-      exists(
-        Flow1::PathNode midEntry1, Flow2::PathNode midEntry2, Flow1::PathNode midExit1,
-        Flow2::PathNode midExit2
+      exists(Flow1::PathNode pred1, Flow2::PathNode pred2 |
+        fwdReachableInterprocEntry(pred1, pred2) and
+        isSuccessor(pred1, pred2, node1, node2)
+      )
+    }
+
+    bindingset[pred1, pred2]
+    bindingset[succ1, succ2]
+    private predicate isSuccessor(
+      Flow1::PathNode pred1, Flow2::PathNode pred2, Flow1::PathNode succ1, Flow2::PathNode succ2
+    ) {
+      exists(Flow1::PathNode mid1, Flow2::PathNode mid2 |
+        localPathStep1*(pred1, mid1) and
+        localPathStep2*(pred2, mid2)
       |
-        reachableInterprocEntry(source1, source2, midEntry1, midEntry2) and
-        interprocEdgePair(midExit1, midExit2, node1, node2) and
-        localPathStep1*(midEntry1, midExit1) and
-        localPathStep2*(midEntry2, midExit2)
+        isSinkPair(mid1, mid2) and
+        succ1 = mid1 and
+        succ2 = mid2
+        or
+        interprocEdgePair(mid1, mid2, succ1, succ2)
       )
     }
 
+    pragma[assume_small_delta]
+    pragma[nomagic]
+    private predicate revReachableInterprocEntry(Flow1::PathNode node1, Flow2::PathNode node2) {
+      fwdReachableInterprocEntry(node1, node2) and
+      (
+        isSinkPair(node1, node2)
+        or
+        exists(Flow1::PathNode succ1, Flow2::PathNode succ2 |
+          revReachableInterprocEntry(succ1, succ2) and
+          isSuccessor(node1, node2, succ1, succ2)
```

Comment:
Replacing this with `fwdIsSuccessor` means that `fwdReachableInterprocEntry(node1, node2)` is implied in this case, so it only needs to be joined in the base case.

Reply:
Good catch! Fixed in 2458fa0ab3aa598306d4aad57fc0809d31dbc7d9.